File: 0701-0800/0785-Is Graph Bipartite/0785-Is Graph Bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        visited = [0] * len(graph)
        def dfs(curr, prev):
            visited[curr] = 1 if prev == -1 else 3 - visited[prev]
            for next in graph[curr]:
                if visited[next] == 0 and not dfs(next, curr):
                    return False
                elif visited[next] == visited[curr]:
                    return False
            return True
        
        for i in range(len(graph)):
            if visited[i] == 0 and not dfs(i, -1):
                return False
        return True

File: 0701-0800/0785-Is Graph Bipartite/0785-Is Graph Bipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> visited(n);
        for (int i = 0; i < n; ++i) {
            if (visited[i] == 0) {
                if (!dfs(graph, i, -1, visited)) {
                    return false;
                }
            }
        }
        return true;
    }
    
private:
    bool dfs(vector<vector<int>>& graph, int curr, int prev, vector<int>& visited) {
        visited[curr] = (prev == -1) ? 1 : 3 - visited[prev];
        for (int next : graph[curr]) {
            if (visited[next] == 0) {
                visited[next] = 3 - visited[curr];
                if (!dfs(graph, next, curr, visited)) {
                    return false;
                }
            }
            else if (visited[next] == visited[curr]) {
                return false;
            }
        }
        
        return true;
    }
};
